### PR TITLE
Part 2 of scraping categories

### PR DESCRIPTION
**Pull request by dshcheung**

### CODE REVIEW

File: lib/tasks/allrecipe-category-tags.rake [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    puts main_category_name
+    puts main_category_url
+    browser = open(main_category_url).read
+    html_doc = Nokogiri::HTML(browser)
+
+    sub_category_elements = html_doc.css('a#hlSubNavItem')
+    if sub_category_elements.any?
+      sub_category_elements.each do |sub_category_element|
+        #prevent inconsistent href links
+        sub_category_url = sub_category_element.attr("href") + "&vm=l"
```
Comment (h8rry) [inline lib/tasks/allrecipe-category-tags.rake:22]: Why do we need `+ "&vm=l"`? Kimono couldn't identify?

File: lib/tasks/allrecipe-category-tags.rake [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        if sub_category_url[/(http:\/\/allrecipes.com)/] == nil
+          sub_category_url = "http://allrecipes.com" + sub_category_url
+        end
+        if sub_category_url == main_category_url
+          next
+        end
+        create_category(main_category_name, sub_category_element.text, sub_category_url)
+        get_category_info(sub_category_element.text, sub_category_url)
```
Comment (h8rry) [inline lib/tasks/allrecipe-category-tags.rake:30]: How many levels are there for this recursion?

File: lib/tasks/allrecipe-category-tags.rake (hunk 1 above)
Comment (dshcheung) [inline lib/tasks/allrecipe-category-tags.rake:22]: vm=1 changes to listing method from grid to list. 
If I don't change to list, I couldn't get the author information
The default link generates the grid system. so i have to put in the parameter to change it

File: lib/tasks/allrecipe-category-tags.rake (hunk 2 above)
Comment (dshcheung) [inline lib/tasks/allrecipe-category-tags.rake:30]: wow....5? 10? 20? i don't know. but i made sure the code doesn't go into infinite loop
i tried scraping everything yesterday and it worked fine

File: lib/tasks/allrecipe-init.rake [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      #gather all information into arrays
+      descriptions = html_doc.css('div > div > div > div > span.popReview')
+      url_code_elements = html_doc.css('div > div > div > h3 > a')
+      authors = html_doc.css('div.searchResult.hub-list-view > div.search_mostPop > div.searchRtSd > div.search_mostPop_Review > span:nth-child(2)')
+
+      #loop each author and check if OutsideProfile exists, if not, create one and create new_recipe
+      authors.each_with_index do |author, index|
+        #check if OutsideProfile Exists, if not, create one and continue
+        if author.css('a').any?
+          author_name = author.css('a').text.squish
```
Comment (h8rry) [inline lib/tasks/allrecipe-init.rake:27]: Nice use of `squish`

File: lib/tasks/allrecipe-init.rake [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        author_existence = OutsideProfile.find_by(username: author_name, outside_profile_url: outside_profile_url)
+        if author_existence == nil
+          author_existence = OutsideProfile.create(display_format: 1, username: author_name, outside_profile_url: outside_profile_url)
+        end
+
+        #extract information and putting it in new_recipe
+        url = url_code_elements[index].attr('href')
+        url_array = url.split('/')
+
+        if Recipe.find_by(url_code: url_array[4], domain_name_id: 1) == nil
```
Comment (h8rry) [inline lib/tasks/allrecipe-init.rake]: Should put `url_array[4]` into a variable since you need to use it later one as well. Easier to understand what `url_array[4]` is

File: lib/tasks/allrecipe-init.rake [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+          #create a new recipe entry
+          new_recipe = author_existence.recipes.new
+
+          new_recipe.outside_profile_id = author_existence.id
+          new_recipe.domain_name_id = 1.to_i
```
Comment (h8rry) [inline lib/tasks/allrecipe-init.rake]: Why `1.to_i`?

File: lib/tasks/allrecipe-init.rake [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+          new_recipe.url_code = url_array[4]
+          new_recipe.name = url_code_elements[index].text
+          new_recipe.description = descriptions[index].text
+
+          #go to part two of scrapping process to scrape recipe in depth information
+          get_recipe_indepth_info(url, new_recipe)
```
Comment (h8rry) [inline lib/tasks/allrecipe-init.rake]: I don't like how `new_recipe.save` is nested inside another function. It's so hard to follow and find. Maybe you can save general information first, and then you have separate task that will scrape in-depth information. It will be easier for you to manage scraping.

File: lib/tasks/allrecipe-init.rake [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    ingredients_elements = html_doc.css('#liIngredient')
+    ingredients_elements.each do |element|
+      #extract allrecipe_ingredient_code
+      #find or create ar_ingredient_code entry and create ingredient_sub_name
+      ar_ingredient_code = element.attr('data-ingredientid').to_i
+      ar_ingredient_code_existence = find_create_ingredient(ar_ingredient_code)
+
+      #extract ingredient_sub_name and create entry
+      ingredient_sub_name = element.css('#lblIngName').text
+      ar_ingredient_code_existence.ingredient_names.create(ingredient_sub_name: ingredient_sub_name)
```
Comment (harrychenca) [inline lib/tasks/allrecipe-init.rake]: I think if you are using functions to create records, then you should stick with your convention. For example, you used `find_create_ingredient`, but then you used `ar_ingredient_code_existence.ingredient_names.create(ingredient_sub_name: ingredient_sub_name)` immediately afterwards.

File: lib/tasks/allrecipe-init.rake [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    #extract instructions info
+    instructions_elements = html_doc.css('div > div > ol > li > span')
+    instructions_array = []
+    instructions_elements.each do |elements|
+      instructions_array.push(elements.text)
+    end
+    new_recipe.instructions = instructions_array
+
+    #save new_recipe
+    new_recipe.save
```
Comment (h8rry) [inline lib/tasks/allrecipe-init.rake]: Hard to follow. Codes should be easy to follow and understand.

File: lib/tasks/allrecipe-init.rake [hunk 9]
```diff
@@ -0,0 +1,172 @@
+task :init, [:start_frequency] => :environment do |task, args|
```
Comment (h8rry) [inline lib/tasks/allrecipe-init.rake]: Can you `desc` describe the task?

File: lib/tasks/allrecipe-recipe_category_relations.rake [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+      when 1
+        retry
+      when 2
+        next
+      end
+    end
+  end
+end
+
+def get_recipe_basic_info(url, category_id)
```
Comment (h8rry) [inline lib/tasks/allrecipe-recipe_category_relations.rake]: How is this different from `init`? Here we scrape basic info, but there we scraped basic and in-depth?

File: lib/tasks/allrecipe-init.rake (hunk 5 above)
Comment (dshcheung) [inline lib/tasks/allrecipe-init.rake]: for some reason it became a string("1") last time i tried it. so i use to_i and the it didn't turn to string.

File: lib/tasks/allrecipe-recipe_category_relations.rake [hunk 11]
```diff
@@ -0,0 +1,73 @@
+desc "scrape the relationship between recipes and a category"
```
Comment (h8rry) [inline lib/tasks/allrecipe-recipe_category_relations.rake:1]: Wouldn't it be a lot faster to associate category with recipes when you were creating recipes?
